include/CL/sycl/math.hpp

```diff
@@ -167,6 +170,129 @@ TRISYCL_MATH_WRAP(trunc)
 //*TRISYCL_MATH_WRAP3(mad24)
 //*TRISYCL_MATH_WRAP3(mul24)
 
+// Note: A lot of these math functions could be forwarded onto the OpenCL
+// implementations when compiling for OpenCL
+
+// Returns the cross product of p0.xyz and p1.xyz. The w component of float4
+// result returned will be 0.0.
+template <typename T, int size>
+vec<T, size> cross(vec<T, size> p0,
+                   vec<T, size> p1) {
+  static_assert(!(size > 4 || size < 3), "vec type should be 3 or 4 dimensions");
+  if constexpr (size > 3) {
+    return vec<T, size> {
+            (p0[1] * p1[2]) - (p0[2] * p1[1]),
+            (p0[2] * p1[0]) - (p0[0] * p1[2]),
+            (p0[0] * p1[1]) - (p0[1] * p1[0]),
+            0.0};
+  } else {
+    return vec<T, size> {
+            (p0[1] * p1[2]) - (p0[2] * p1[1]),
+            (p0[2] * p1[0]) - (p0[0] * p1[2]),
+            (p0[0] * p1[1]) - (p0[1] * p1[0])};
+  }
+}
+
+// Return the length of vector p, i.e., sqrt(p.x^2 + p.y^2 + ...)
+template <typename T, int size>
+T length(vec<T, size> p) {
+  auto temp = p * p;
+  return std::sqrt(std::accumulate(temp.begin(), temp.end(), 0.0));
```

Comment:
Why is it `0.0` and not `T {}` for example or similar to the `dot` one?

Otherwise you are forcing the computation to `double` and then cast back for `float`...